packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -0,0 +1,358 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math' as math;
+
+import 'box.dart';
+import 'object.dart';
+
+enum WrapAlignment {
+  start,
+  center,
+  end,
+  // TODO(ianh): justify
+}
+
+class WrapParentData extends ContainerBoxParentDataMixin<RenderBox> { }
+
+class RenderWrap extends RenderBox with ContainerRenderObjectMixin<RenderBox, WrapParentData>,
+                                        RenderBoxContainerDefaultsMixin<RenderBox, WrapParentData> {
+  RenderWrap({
+    List<RenderBox> children,
+    Axis direction: Axis.horizontal,
+    WrapAlignment mainAxisAlignment: WrapAlignment.center,
+    WrapAlignment crossAxisAlignment: WrapAlignment.center,
+    double spacing,
+  }) : _direction = direction,
+       _mainAxisAlignment = mainAxisAlignment,
+       _crossAxisAlignment = crossAxisAlignment,
+       _spacing = spacing {
+    assert(direction != null);
+    assert(mainAxisAlignment != null);
+    assert(crossAxisAlignment != null);
+    assert(spacing != null);
+    addAll(children);
+  }
+
+  /// The direction to use as the main axis.
+  // TODO(ianh): we should also support growing down vs growing up and growing left vs growing right
+  Axis get direction => _direction;
+  Axis _direction;
+  set direction (Axis value) {
+    assert(value != null);
+    if (_direction != value) {
+      _direction = value;
+      markNeedsLayout();
+    }
+  }
+
+  /// How the children should be placed along the main axis.
+  WrapAlignment get mainAxisAlignment => _mainAxisAlignment;
+  WrapAlignment _mainAxisAlignment;
+  set mainAxisAlignment (WrapAlignment value) {
+    assert(value != null);
+    if (_mainAxisAlignment != value) {
+      _mainAxisAlignment = value;
+      markNeedsLayout();
+    }
+  }
+
+  /// How the children should be placed along the cross axis.
+  WrapAlignment get crossAxisAlignment => _crossAxisAlignment;
+  WrapAlignment _crossAxisAlignment;
+  set crossAxisAlignment (WrapAlignment value) {
+    assert(value != null);
+    if (_crossAxisAlignment != value) {
+      _crossAxisAlignment = value;
+      markNeedsLayout();
+    }
+  }
+
+  double get spacing => _spacing;
```

Comment:
mainSpacing/crossSpacing?

I fear the main/cross distinction is a bit confused here because things grow in both directions...